62_CAD/shaders/main_pipeline/fragment_shader.hlsl

```diff
@@ -1,655 +1,1104 @@
+                transitionInfo.boundaryHeight = currentIntervalHeight;
+            }
+            else
+            {
+                int nextIntervalIdx = intervalIndex + 1;
+                getIntervalHeightAndColor(nextIntervalIdx, settings, transitionInfo.otherSegmentColor, transitionInfo.boundaryHeight);
+            }
+            
+            float4 localHeightColor = smoothHeightSegmentTransition(transitionInfo, heightDeriv);
+            outputColor.rgb = localHeightColor.rgb;
+            outputColor.a *= localHeightColor.a;
+        }
+        else if (mode == E_HEIGHT_SHADING_MODE::CONTINOUS_INTERVALS)
+        {
+            DTMSettingsHeightsAccessor dtmHeightsAccessor = { settings };
+            uint32_t upperBoundHeightIndex = nbl::hlsl::upper_bound(dtmHeightsAccessor, 0, heightMapSize, height);
+            uint32_t lowerBoundHeightIndex = upperBoundHeightIndex == 0 ? upperBoundHeightIndex : upperBoundHeightIndex - 1;
+
+            float upperBoundHeight = settings.heightColorMapHeights[upperBoundHeightIndex];
+            float lowerBoundHeight = settings.heightColorMapHeights[lowerBoundHeightIndex];
+
+            float4 upperBoundColor = settings.heightColorMapColors[upperBoundHeightIndex];
+            float4 lowerBoundColor = settings.heightColorMapColors[lowerBoundHeightIndex];
+
+            float interpolationVal;
+            if (upperBoundHeightIndex == 0)
+                interpolationVal = 1.0f;
+            else
+                interpolationVal = (height - lowerBoundHeight) / (upperBoundHeight - lowerBoundHeight);
+
+            float4 localHeightColor = lerp(lowerBoundColor, upperBoundColor, interpolationVal);
+
+            outputColor.a *= localHeightColor.a;
+            outputColor.rgb = localHeightColor.rgb * outputColor.a + outputColor.rgb * (1.0f - outputColor.a);
+        }
+    }
+
+    return outputColor; 
+}
+
+float4 calculateDTMContourColor(in DTMContourSettings contourSettings, in float3 v[3], in uint2 edgePoints[3], in float2 fragPos, in float height)
+{
+    float4 outputColor;
+
+    LineStyle contourStyle = loadLineStyle(contourSettings.contourLineStyleIdx);
+    const float contourThickness = (contourStyle.screenSpaceLineWidth + contourStyle.worldSpaceLineWidth * globals.screenToWorldRatio) * 0.5f;
+    float stretch = 1.0f;
+    float phaseShift = 0.0f;
+
+    // TODO: move to ubo or push constants
+    const float startHeight = contourSettings.contourLinesStartHeight;
+    const float endHeight = contourSettings.contourLinesEndHeight;
+    const float interval = contourSettings.contourLinesHeightInterval;
+
+    // TODO: can be precomputed
+    const int maxContourLineIdx = (endHeight - startHeight + 1) / interval;
+
+    // TODO: it actually can output a negative number, fix
+    int contourLineIdx = nbl::hlsl::_static_cast<int>((height - startHeight + (interval * 0.5f)) / interval);

```

Comment:
(interval*0.5f)/interval is just 0.5f, 
you can just do `(height - startHeight) / interval + 0.5f`